lucene/benchmark/src/java/org/apache/lucene/benchmark/byTask/feeds/DirContentSource.java

```diff
@@ -0,0 +1,270 @@
+        List<Path> found = new ArrayList<>();
+        for (Path p : stream) {
+          if (Files.isDirectory(p)) {
+            found.add(p);
+          }
+        }
+        push(found.toArray(new Path[found.size()]));
+      }
+      try (DirectoryStream<Path> stream = Files.newDirectoryStream(f, "*.txt")) {
+        List<Path> found = new ArrayList<>();
+        for (Path p : stream) {
+          found.add(p);
+        }
+        push(found.toArray(new Path[found.size()]));
+      }
+      find();
+    }
+
+    void push(Path[] files) {
+      Arrays.sort(files, c);
+      for(int i = 0; i < files.length; i++) {
+        // System.err.println("push " + files[i]);
+        stack.push(files[i]);
+      }
+    }
+
+    public int getCount(){
+      return count;
+    }
+
+    @Override
+    public boolean hasNext() {
+      return stack.size() > 0;
+    }
+    
+    @Override
+    public Path next() {
+      assert hasNext();
+      count++;
+      Path object = stack.pop();
+      // System.err.println("pop " + object);
+      try {
+        find();
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+      return object;
+    }
+
+    @Override
+    public void remove() {
+      throw new RuntimeException("cannot");
+    }
+
+  }
+  
+  private ThreadLocal<DateFormatInfo> dateFormat = new ThreadLocal<>();
+  private Path dataDir = null;
+  private int iteration = 0;
+  private Iterator inputFiles = null;
+
+  // get/initiate a thread-local simple date format (must do so 
+  // because SimpleDateFormat is not thread-safe).
+  private DateFormatInfo getDateFormatInfo() {
+    DateFormatInfo dfi = dateFormat.get();
+    if (dfi == null) {
+      dfi = new DateFormatInfo();
+      dfi.pos = new ParsePosition(0);
+      // date format: 30-MAR-1987 14:22:36.87
+      dfi.df = new SimpleDateFormat("dd-MMM-yyyy kk:mm:ss.SSS", Locale.ENGLISH);
+      dfi.df.setLenient(true);
+      dateFormat.set(dfi);
+    }
+    return dfi;
+  }
+  
+  private Date parseDate(String dateStr) {
+    DateFormatInfo dfi = getDateFormatInfo();
+    dfi.pos.setIndex(0);
+    dfi.pos.setErrorIndex(-1);
+    return dfi.df.parse(dateStr.trim(), dfi.pos);
+  }
+
+  @Override
+  public void close() throws IOException {
+    inputFiles = null;
+  }
+  
+  @Override
+  public DocData getNextDocData(DocData docData) throws NoMoreDataException, IOException {
+    Path f = null;
+    String name = null;
+    synchronized (this) {
+      if (!inputFiles.hasNext()) { 
+        // exhausted files, start a new round, unless forever set to false.
+        if (!forever) {
+          throw new NoMoreDataException();
+        }
+        inputFiles = new Iterator(dataDir);
+        iteration++;
+      }
+      f = inputFiles.next();
+      // System.err.println(f);
+      name = f.toRealPath()+"_"+iteration;
+    }
+    
+    BufferedReader reader = Files.newBufferedReader(f, StandardCharsets.UTF_8);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *reader*. The resource is referenced in statements at the following lines: 220, 221, 222, 223, 225. The resource closure statement is at line: 228. There are other execution paths that do not contain closure statements, e.g., when exception is thrown by BufferedReader.readLine. Either a) close *reader* in a try-finally block or b) close the resource by declaring *reader* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).